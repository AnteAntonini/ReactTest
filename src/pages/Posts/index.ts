import { lazy } from "react";

export const Posts = lazy(() => import("./Posts"));
