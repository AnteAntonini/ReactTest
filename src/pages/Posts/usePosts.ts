import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { useFetch } from "../../hooks/useFetch";
import { TUsersResponse } from "../../types/TUsers";
import { TPostsCommentsResponse } from "../../types/TComments";
import { TPostsResponse, TPostsWithComentsAndUsers } from "../../types/TPosts";

export const usePosts = (message: string) => {
  const [filteredPosts, setFilteredPosts] =
    useState<TPostsWithComentsAndUsers[]>();
  const [searchUsername, setSearchUsername] = useState("");
  const navigate = useNavigate();

  let commentsByPostId: TPostsCommentsResponse[][] = [];

  useEffect(() => {
    console.log(`${message} Posts`);
  }, []);

  const users: TUsersResponse[] = useFetch(
    "https://demo.martian.services/api/users"
  );

  const comments: TPostsCommentsResponse[] = useFetch(
    "https://demo.martian.services/api/comments"
  );
  const posts: TPostsResponse[] = useFetch(
    "https://demo.martian.services/api/posts"
  );

  comments &&
    comments.forEach((comment) => {
      const postId = comment.postId;
      if (!commentsByPostId[postId]) {
        commentsByPostId[postId] = [];
      }
      commentsByPostId[postId].push(comment);
    });

  const PostsWithComentsAndUsers =
    posts &&
    posts.map((post) => {
      const postId = post && post.id;
      let mappedUser = { name: "" };

      const postUser: TUsersResponse | undefined =
        users && users.find((user) => user.id === post.userId);
      const postComments = commentsByPostId[postId] || [];

      if (postUser) {
        mappedUser.name = postUser.name;
      }

      return Object.assign({}, post, mappedUser, {
        comments: postComments,
      });
    });

  const findUserByName = (username: string) => {
    return users.find((user) =>
      user.name.toLowerCase().includes(username.toLowerCase())
    );
  };

  useEffect(() => {
    const filterPosts = () => {
      const originalPosts = PostsWithComentsAndUsers || [];

      if (searchUsername.trim() === "") {
        setFilteredPosts(originalPosts);
      } else {
        const matchedUser = findUserByName(searchUsername);

        if (matchedUser) {
          const matchedPosts = originalPosts.filter(
            (post) => post.userId === matchedUser.id
          );
          setFilteredPosts(matchedPosts);
        } else {
          setFilteredPosts([]);
        }
      }
    };

    filterPosts();
  }, [searchUsername, users]);

  const handleClick = (postId: number) => {
    navigate(`/post/${postId}`);
  };

  const onChange = (e: any) => setSearchUsername(e.target.value);

  return {
    filteredPosts,
    onChange,
    handleClick,
  };
};
