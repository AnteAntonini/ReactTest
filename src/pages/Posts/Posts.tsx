import { usePosts } from "./usePosts";
import SearchInput from "../../components/searchInput/SearchInput";
import PostItem from "../../components/PostItem/PostItem";
import { TPostsWithComentsAndUsers } from "../../types/TPosts";

const Posts = ({ message }: { message: string }) => {
  const { filteredPosts, onChange, handleClick } = usePosts(message);

  return (
    <div className="flex-col-center">
      <h1 className="page-title">Posts</h1>
      <SearchInput onChange={onChange} message="Hello from" />
      {filteredPosts &&
        filteredPosts.map((post: TPostsWithComentsAndUsers) => (
          <PostItem
            key={post.id}
            id={post.id}
            onClick={() => handleClick(post.id)}
            postTitle={post.title}
            postUserName={post.name}
            comments={post.comments}
            message="Hello from"
          />
        ))}
    </div>
  );
};

export default Posts;
