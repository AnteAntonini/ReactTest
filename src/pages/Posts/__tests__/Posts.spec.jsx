import { render, fireEvent, screen } from "@testing-library/react";
import Posts from "../Posts";
import { usePosts } from "../usePosts";

// Mock the usePosts hook
jest.mock("../usePosts");

describe("Posts Component", () => {
  it("renders posts with search input", () => {
    // Mock the usePosts hook return values
    usePosts.mockReturnValue({
      filteredPosts: [
        {
          id: 1,
          title: "Sample Post 1",
          name: "User1",
          comments: [],
        },
        {
          id: 2,
          title: "Sample Post 2",
          name: "User2",
          comments: [],
        },
      ],
      onChange: jest.fn(),
      handleClick: jest.fn(),
    });

    render(<Posts message="Test" />);

    // Check if the page title, search input, and post items are present
    expect(screen.getByText("Posts")).toBeInTheDocument();
    expect(screen.getByTestId("search-input")).toBeInTheDocument();
    expect(screen.getByText("Sample Post 1")).toBeInTheDocument();
    expect(screen.getByText("User1")).toBeInTheDocument();
    expect(screen.getByText("Sample Post 2")).toBeInTheDocument();
    expect(screen.getByText("User2")).toBeInTheDocument();
  });

  it("calls handleClick when a post item is clicked", () => {
    // Mock the usePosts hook return values
    const mockHandleClick = jest.fn();
    usePosts.mockReturnValue({
      filteredPosts: [
        {
          id: 3,
          title: "Sample Post 3",
          name: "User3",
          comments: [],
        },
      ],
      onChange: jest.fn(),
      handleClick: mockHandleClick,
    });

    render(<Posts message="Test" />);

    const postItem = screen.getByText("Sample Post 3");

    // Simulate a click on the post item
    fireEvent.click(postItem);

    // Verify that handleClick is called with the correct arguments
    expect(mockHandleClick).toHaveBeenCalledWith(3);
  });
});

describe("Test snapshot", () => {
  it("Component should match snapshot", () => {
    usePosts.mockReturnValue({
      filteredPosts: [
        {
          id: 1,
          title: "Sample Post 1",
          name: "User1",
          comments: [],
        },
        {
          id: 2,
          title: "Sample Post 2",
          name: "User2",
          comments: [],
        },
      ],
      onChange: jest.fn(),
      handleClick: jest.fn(),
    });
    const { asFragment } = render(<Posts message="Test" />);

    expect(asFragment()).toMatchSnapshot();
  });
});
