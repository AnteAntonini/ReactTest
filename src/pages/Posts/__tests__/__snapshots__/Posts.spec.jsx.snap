// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex-col-center"
  >
    <h1
      class="page-title"
    >
      Posts
    </h1>
    <div
      class="row"
    >
      <input
        class="search"
        data-testid="search-input"
        placeholder="Search..."
        type="text"
      />
    </div>
    <div
      class="post-section post-hover"
    >
      <div
        class="post-title"
      >
        Sample Post 1
      </div>
      <span
        class="post-username"
      >
        User1
      </span>
      <hr />
      <div>
        <h3
          class="comment-title"
        >
           0 Comments
        </h3>
      </div>
    </div>
    <div
      class="post-section post-hover"
    >
      <div
        class="post-title"
      >
        Sample Post 2
      </div>
      <span
        class="post-username"
      >
        User2
      </span>
      <hr />
      <div>
        <h3
          class="comment-title"
        >
           0 Comments
        </h3>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
